addons/l10n_in_ewaybill_stock/tools/ewaybill_api.py

```diff
@@ -0,0 +1,207 @@
+from odoo.addons.l10n_in_edi_ewaybill.models.error_codes import ERROR_CODES
+
+
+_logger = logging.getLogger(__name__)
+
+
+class EWayBillError(Exception):
+
+    def __init__(self, response):
+        self.error_json = self._set_missing_error_message(response)
+        self.error_json.setdefault('odoo_warning', [])
+        self.error_codes = self.get_error_codes()
+        super().__init__(response)
+
+    def _set_missing_error_message(self, response):
+        for error in response.get('error', []):
+            if error.get('code') and not error.get('message'):
+                error['message'] = self._find_missing_error_message(error.get('code'))
+        return response
+
+    @staticmethod
+    def _find_missing_error_message(code):
+        error_message = ERROR_CODES.get(code)
+        return error_message or _("We don't know the error message for this error code. Please contact support.")
+
+    def get_all_error_message(self):
+        return Markup("<br/>").join(
+            ["[%s] %s" % (e.get("code"), e.get("message")) for e in self.error_json.get('error')]
+        )
+
+    def get_error_codes(self):
+        return [e.get("code") for e in self.error_json['error']]
+
+
+class EWayBillApi:
+
+    DEFAULT_HELP_MESSAGE = _(
+        "Somehow this E-waybill has been %s in the government portal before. "
+        "You can verify by checking the details into the government "
+        "(https://ewaybillgst.gov.in/Others/EBPrintnew.asp)"
+    )
+
+    def __init__(self, company):
+        company.ensure_one()
+        self.company = company
+        self.env = self.company.env
+
+    def _ewaybill_connect_to_server(self, url_path, params):
+        user_token = self.env["iap.account"].get("l10n_in_edi")
+        params.update({
+            "account_token": user_token.account_token,
+            "dbuuid": self.env["ir.config_parameter"].sudo().get_param("database.uuid"),
+            "username": self.company.sudo().l10n_in_edi_ewaybill_username,
+            "gstin": self.company.vat,
+        })
+        if self.company.sudo().l10n_in_edi_production_env:
+            default_endpoint = DEFAULT_IAP_ENDPOINT
+        else:
+            default_endpoint = DEFAULT_IAP_TEST_ENDPOINT
+        endpoint = self.env["ir.config_parameter"].sudo().get_param("l10n_in_edi_ewaybill.endpoint", default_endpoint)
+        url = f"{endpoint}{url_path}"
+        try:
+            response = jsonrpc(url, params=params, timeout=70)
+            if response.get('error'):
+                raise EWayBillError(response)
+        except AccessError as e:
+            _logger.warning("Connection error: %s", e.args[0])
+            return {
+                "error": [{
+                    "code": "access_error",
+                    "message": _("Unable to connect to the E-WayBill service."
+                                 "The web service may be temporary down. Please try again in a moment.")
+                }]
+            }
```

Comment:
Why are we raising an Exception in the error in the response case and here we return the dict?  (we should use the same mechanism for both unless there is some logic in which they are treated differently that would make sense...)